File: js/src/engine/texture-manager.js
'use strict';

import TexturePage from './texture-page';

const [REFERENCES, PAGES, SOURCES] = [Symbol(), Symbol(), Symbol()];

class TextureManager {
  [REFERENCES] = new WeakMap();
  [PAGES] = {};
  [SOURCES] = null;

  constructor(sources = []) {
    this[SOURCES] = sources;
  }

  // load a set of texture pages
  load(textures) {
    textures = new Set(textures);
    const loaded = [];
    for(let texture of textures) {
      if(!this[PAGES][texture]) {
        this[PAGES][texture] = new TexturePage(this[SOURCES][texture]);
        loaded.push(this[PAGES][texture].loaded);
      }
      this[REFERENCES].set(
        this[PAGES][texture],
        (this[REFERENCES].get(this[PAGES][texture]) || 0) + 1
      );
    }
    return Promise.all(loaded);
  }

  // remove all texture pages in the list given
  purge(textures = []) {
    for(let texture of textures) {
      if(this[PAGES][texture]) {
        this[REFERENCES].set(
          this[PAGES][texture],
          this[REFERENCES].get(this[PAGES][texture]) - 1
        );
        if(!this[REFERENCES].get(this[PAGES][texture])) {
          delete this[PAGES][texture];
        }
      }
    }
  }

  get pages() { return this[PAGES]; }

  // instantiate the sprite from the pages
  sprite(sprite) {
    let spr;
    for(let page of Object.keys(this[PAGES])) {
      if(spr = this[PAGES][page].make(sprite)) {
        break;
      }
    }
    if(!spr) { throw `Sprite ${sprite} does not exist in the current set of texture pages`; }
    return spr;
  }
}

export default TextureManager;

File: c++/src/engine/texture-manager.cc
#include "texture-manager.h"
#include "sprite.h"

namespace Game {
    TextureManager::TextureManager(SDL_Renderer & ren, std::map<std::string, std::string> & sources)
        : _sources{ sources }, _ren{ ren } {}

    void TextureManager::load(const std::vector<std::string> & textures) {
        for(const std::string & name : textures) {
            if(!_pages.count(name)) {
                try {
                    _pages[name] = std::make_unique<TexturePage>(_sources[name], _ren);
                } catch( std::string err ) {
                    throw;
                } catch( ... ) {
                    throw "Could not load texture page " + name;
                }
            }
            ++_references[name];
        }
    }

    void TextureManager::purge(const std::vector<std::string> & textures) {
        for(const std::string & name : textures) {
            if(_references[name]) {
                if(!--_references[name]) {
                    _pages.erase(name);
                }
            }
        }
    }

    std::unique_ptr<Sprite> TextureManager::sprite(const std::string & name) {
        for(auto & page : _pages) {
            if(page.second->hasSprite(name)) {
                return page.second->make(name);
            }
        }
        throw "Sprite " + name + " does not exist in the current set of texture pages!s";
    }

    TexturePage & TextureManager::texture(const std::string & name) {
        if(!_pages.count(name)) {
            throw "No texture page named " + name + " exists";
        }
        return *_pages[name];
    }
}
